modules/pam_setquota/pam_setquota.c

```diff
@@ -0,0 +1,387 @@
+/* PAM setquota module
+
+   This PAM module sets disk quota when a session begins.
+
+   Copyright © 2006 Ruslan Savchenko <savrus@mexmat.net>
+   Copyright © 2010 Shane Tzen <shane@ict.usc.edu>
+   Copyright © 2012-2020 Sven Hartge <sven@svenhartge.de>
+   Copyright © 2016 Keller Fuchs <kellerfuchs@hashbang.sh>
+
+*/
+
+#include <sys/types.h>
+#include <sys/quota.h>
+#include <linux/quota.h>
+#include <pwd.h>
+#include <syslog.h>
+#include <ctype.h>
+#include <errno.h>
+#include <mntent.h>
+#include <stdio.h>
+#include <stdbool.h>
+
+#define PAM_SM_SESSION
+
+#include <security/pam_modules.h>
+#include <security/_pam_macros.h>
+#include <security/pam_ext.h>
+#include <security/pam_modutil.h>
+#include "pam_inline.h"
+
+#ifndef PATH_LOGIN_DEFS
+# define PATH_LOGIN_DEFS "/etc/login.defs"
+#endif
+
+#define MAX_UID_VALUE 0xFFFFFFFFUL
+
+struct pam_params {
+  uid_t start_uid;
+  uid_t end_uid;
+  const char *fs;
+  size_t fs_len;
+  int overwrite;
+  int debug;
+};
+
+static inline void
+debug(pam_handle_t *pamh, struct if_dqblk *p,
+      const char *device, const char *dbgprefix) {
+  pam_syslog(pamh, LOG_DEBUG, "%s device=%s bsoftlimit=%llu bhardlimit=%llu "
+                              "isoftlimit=%llu ihardlimit=%llu",
+             dbgprefix, device,
+             p->dqb_bsoftlimit, p->dqb_bhardlimit,
+             p->dqb_isoftlimit, p->dqb_ihardlimit);
+}
+
+static unsigned long long
+str_to_dqb_num(const char *str) {
+  char *ep = NULL;
+
+  errno = 0;
+  long long temp = strtoll(str, &ep, 10);
+  if (!ep || errno != 0 || (temp < 0))
+    return 0;
```

Comment:
This implements roughly the following logic: if the given string is not a number from our PoV, ignore it by treating it as 0 (which means the corresponding quota limit is disabled).  Should we issue some diagnostics in this case?

Reply:
We could. But that would mean dragging `pamh` through both `parse_dqblk` and `str_to_dqb_num` to be able to use `pam_syslog()`, wouldn't it?

Reply:
Yes, it would.

Reply:
I will do the same for `str_to_uid()` then.

Reply:
I have added some more output on parse errors, according to Linux-PAM_MWG at the LOG_ERR level.